### ISSUE TITLE
Aircraft readback affirmative in response to dvs but do not actually descend

### ISSUE THREAD

**Issue by cake-pie** [openscope/openscope#1734]
via https://github.com/openscope/openscope/issues/1455#issuecomment-769087246

Airport: LOWW, ENGM, possibly others

### LOWW

Airport is configured with arrivals on RWY 34. When issued `dvs` all aircraft respond with readback "descend via STAR and maintain 3000". This is due to corresponding transitions having fix `WW971` with `A030+`. If a specific altitude is supplied as an argument e.g. `dvs 40` or  `dvs 20` aircraft will readback with the given altitude. However, in all cases, aircraft **do not descend** despite the affirmative readback.

In the LOWW airport file, most fixes altitude constraints along STARs and transitions are of "at or above" `Axxx+` type. This is in accordance with reality ([Austro Control eAIP](https://eaip.austrocontrol.at/)). As @Fechulo explained [here](#issuecomment-621199815) these altitude restrictions are for planning purposes only and actual descent clearances are given by ATC.

Due to absence of "at or below" crossing restrictions, aircraft are technically remaining in compliance with STARs and transitions.

However it is a problem if aircraft respond with readback in the affirmative but never actually begin descent.

### ENGM

Aircraft when issued `dvs` readback with "descend via STAR and maintain 5000", but settle at a final altitude of FL 90, 100 or 110 even though their flight plans all say FL50

As [explained](https://github.com/openscope/openscope/issues/1455#issuecomment-753704875) by @erikquinn, `dvs` without altitude specified tells aircraft to descend as compelled to by STARs, which have crossing restrictions at `A090`, `A100` or `A110` respectively, hence the observed final altitudes. Holding at VALPU and INSUV have "at or above" restriction (`A050+`) and aircraft are thus not compelled to descend to FL50. Airport file is accurate ([Avinor eAIP](https://ais.avinor.no/no/AIP/View/100/history-no-NO.html)).

Same as LOWW, there is an issue of misleading readback.

As for `dvs xx` with altitude specified explicitly, aircraft reply with the given altitude in readback. Actual behavior: if issued `dvs 51` (or above) upon entering the airspace, aircraft *will* descend to FL51 after crossing GM407 / GM455 / GM414 / GM454 at FL 90/100/110. Aircraft will not descend past FL 90/100/110 if issued `dvs 50` or lower, despite affirmative readback.

This is very odd, I was originally expecting aircraft to not descend past FL 90/100/110 at all. It seems the problem isn't quite so simple as "refuse to descend if not compelled by crossing restrictions".

## Steps to reproduce
- try `dvs`, `dvs 40` and `dvs 20` at LOWW
- try `dvs`, `dvs 51` and `dvs 50` at ENGM

## Desired behavior

- aircraft should not reply with an affirmative readback if no descent will actually happen.

- `dvs` without altitude specified should result in "no altitude assigned" readback if not compelled by STAR or transitions to descend from (current? / assigned?) altitude.

- `dvs xxx` with altitude specified should result in descent to eventually reach and maintain the given altitude while complying with crossing restrictions in the STAR
  - fixes with `Ayyy` or `Ayyy+` where `yyy`>`xxx` should prevent descent below `yyy` until past the fix, but should not prevent descent up to `yyy` before reaching the fix

- affirmative readbacks should reflect final altitude that aircraft will maintain if not subsequently issued any further clearance, or compelled by future crossing restrictions.

**Comment by OlaRune** [comment]
Very well written issue! :)

Just one opinion on the desired behaviour: Shouldn't `dvs` without altitude specification automatically make them descend to the final altitude in their flight plan (if they have one) instead of giving the "no altitude assigned" readback?

**Comment by cake-pie** [comment]
Re what the behavior should be, I'm not sure either. I'm a software / coding person and my aviation / ATC knowledge isn't very deep so will defer to the actual ATC people on the finer details.

From a software engineering standpoint, the more questions we ask about various corner cases, the better. It's best to have a full and clear picture of the requirements before actually coding stuff! =)

**Comment by OlaRune** [comment]
> Re what the behavior should be, I'm not sure either. I'm a software / coding person and my aviation / ATC knowledge isn't very deep so will defer to the actual ATC people on the finer details.
> 
> From a software engineering standpoint, the more questions we ask about various corner cases, the better. It's best to have a full and clear picture of the requirements before actually coding stuff! =)

I guess @erikquinn or @Fechulo can answer what the desired behaviour should be.

**Comment by erikquinn** [comment]
Yes and no. It should make them descend only as far as the star forces them if no altitude is specified. Currently they _do_ that, but they show that they are descending to a lower altitude which they never will reach (an issue with `calculateBottomAltitude()`). As far as improving the responses, if there is not enough information in the arrival to descend via, they should be rejecting the instruction because they're unable (and I don't know if that happens or not).

**Comment by Fechulo** [comment]
As I explained [here](https://github.com/openscope/openscope/issues/1455#issuecomment-769442877), part of the problem is the inconsistency across airport files. 

> These altitude restrictions are for planning purposes only and actual descent clearances are given by ATC

I'm not sure I stand by this anymore. It means users need to check every chart to see if dvs will actually do anything or if they need to descend aircraft 'manually'. Even if it's not realistic, I wouldn't be opposed to adding these kinds of restrictions to the STAR definitions, for playability and accessibility's sake

**Comment by OlaRune** [comment]
> As I explained [here](https://github.com/openscope/openscope/issues/1455#issuecomment-769442877), part of the problem is the inconsistency across airport files.
> 
> > These altitude restrictions are for planning purposes only and actual descent clearances are given by ATC
> 
> I'm not sure I stand by this anymore. It means users need to check every chart to see if dvs will actually do anything or if they need to descend aircraft 'manually'. Even if it's not realistic, I wouldn't be opposed to adding these kinds of restrictions to the STAR definitions, for playability and accessibility's sake

But I guess `dvs xx` is still valid at european airports, right? Since this will be a "descent clearance given by ATC", but they will still respect the minimum altitudes in the STAR. Then everything in the "desired behaviour" section above still stands?

**Comment by Fechulo** [comment]
> But I guess dvs xx is still valid at european airports, right?

Not in the real world, which is why it would be slightly unrealistic to allow users to descend aircraft using the DVS command at European airports. I'd like to hear other opinions on this, but I would be willing to sacrifice realism in this case.

Everything in the desired behaviour section sounds good to me

**Comment by cake-pie** [comment]
> [`dvs` without altitude argument] should make them descend only as far as the star forces them [...] Currently they do that, but they show that they are descending to a lower altitude which they never will reach

> if there is not enough information in the arrival to descend via, they should be rejecting the instruction because they're unable (and I don't know if that happens or not)

OK, so for `dvs` with no altitude specified at LOWW, the behavior is correct, and the readback is incorrect, instruction needs to be rejected with "no altitude assigned" which conforms to behavior.

>  (an issue with `calculateBottomAltitude()`)

Is this also responsible for the assigned altitude displayed at middle of second column on aircraft strips? E.g. here are LOWW arrivals showing assigned 30 upon entering the airspace, no commands issued yet:

![loww-arrival-assigned-alt](https://user-images.githubusercontent.com/28502722/106293854-6e917480-6289-11eb-8ad8-a5ce3f2148b8.png)


Is this also incorrect? (The assigned alt on the flight strip probably contributed to why @OlaRune expected "descend to the final altitude in their flight plan".)

---

I get the point that ATC European airports may not use "descend via" phraseology at all and their STARs are not be designed with "descend via" in mind, hence their airport files may "lack" at-or-below altitude restrictions. But that's completely fine! It just means the sim is consistent with reality, rather than these airports somehow being "inconsistent" compared to other airports in the sim. We should not feel compelled to artificially add restrictions that do not exist in the real world, just for the sake of "make `dvs` actually do something".

Instead I think the objective should be: consistent, coherent, logical behavior of `dvs` and `dvs xx` across the board, which is currently not the case -- readback conflicts with actual behavior, nevermind which is correct or not.

My attempt to outline desired behavior is based on a "back to basics" approach: where it is used, the "descend via" phraseology means: comply with all published crossing restrictions, and arrive at some given destination altitude, or if none is specified, descend no lower than is compulsory. That definition is a workable starting point; even if trying to use it with airports and STARs that are not intended, stick to the definition and do the logical thing and we should be fine -- of course, that could entail doing nothing as well, if that is what makes sense.

Question: STARs can come with various combinations of presence or absence of different types of crossing restrictions (`Axxx`/`Axxx+`/`Axxx-`). `Axxx`/`Axxx-` restrictions define upper altitude limits, `Axxx`/`Axxx+` restrictions define lower altitude limits. Where such limits exist, are they guaranteed to be [monotonically](https://en.wikipedia.org/wiki/Monotonic_function) decreasing? Seems like that should be the case, for basic sanity.

From a coder standpoint, I'd like for a robust implementation of `dvs` that is able to accommodate all reasonable combinations, so this is an attempt to clarify the bounds of "reasonable".

---

> allow users to descend aircraft using the DVS command at European airports. I'd like to hear other opinions on this

I feel it's not worth the effort to implement special casing prohibiting certain commands at certain airports. Instead, it should be possible to issue `dvs` and `dvs xx` at any airport. Depending on the STAR, it may or may not do anything, and may or may not be useful.

- `dvs` at LOWW, EGCC, KMIA: without explicit altitude assignment, aircraft don't know what altitude to descend to, readback "unable, no altitude assigned"

- `dvs xx` at EGCC, KMIA: indistinguishable from `a xx`.

- `dvs 35` at LOWW upon entering airspace: aircraft will descend as allowed, i.e. not lower than 6000 until passed WW974, then not lower than 4000 until after WW972, and maintain 3500 (compliant with `A030+` at WW971). That's kind of useful. Compare alternatives:
  - immediate `a 35` would violate the altitude constraints
  - multiple successive manual descent clearance `a xx` as aircraft pass certain points: maybe that's a bit too much babysitting for your liking?
  - you could give multiple `cross` commands in advance, but `dvs xx` a lot less verbose. (`cross` gives you more control of specific altitudes, though.)

**Comment by OlaRune** [comment]
I agree that we should not add artificial constraints to the stars in order to make dvs do something, that will just be confusing and inconsistent. Better then to add an unrealistic behaviour for dvs on such airports or just disable the command so it can't be used at all.

You are correct that the assigned altitude on spawn is what  made me think they would/should strive for that altitude when told to dvs.
